### ISSUE TITLE
C.50: static member function

### ISSUE THREAD

**Issue by tlanc007** [isocpp/CppCoreGuidelines#837]
The example for the factory contains a static member function.  

There are a number of guidelines and examples on static variables. To the best of my recollection this is only use or mention of static member functions in the Guidelines.  For completeness, is it worth having a specific guideline on static member functions?

**Comment by jwakely** [comment]
Guidelines are there when they say something useful, not for completeness. What would it say?

**Comment by tlanc007** [comment]
Yeah, I see your point.

I brought it up because static methods doesn’t come up too much in the literature and I thought some might have it useful to have a definitive explanation of when they could be used, possibly abused and to make a distinction between a static function and a static class function.

As a jumping off point here’s my attempt.

C.X Prefer class functions to static class functions

Static class functions should be use judiciously.  A static class function exists outside the class and can be used without a properly instantiated object of that class.  Before making a class function static, one should consider if the class design is better served leaving it as a normal class function or perhaps even making it global function (#44 Prefer writing nonmember nonfriend functions: C++ Coding standards). 

There are a number of legitimate uses for static class functions.  One such use is to overcome restrictions with constructors.   The Factory design pattern is one such example that takes advantage of this.  

Additionally certain libraries take advantage of static class functions as a form of organization.  For example, limits, from the Standard Library provides the class template numeric_limits:

```
template <class T> class numeric_limits {
public:
  static constexpr T min() noexcept { return T(); }
  static constexpr T max() noexcept { return T(); }
  static constexpr T lowest() noexcept { return T(); }
  ...
};
```

**Comment by MikeGitb** [comment]
What I'm missing is an explanation of why you should avoid them in the first place. In particular, I don't see why normal member functions are preferable to static member functions.

**Comment by cubbimew** [comment]
I think the appropriate rule here already exists: [C.4: Make a function a member only if it needs direct access to the representation of a class](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#Rc-member) , it just needs a note that it applies to both static and non-static members.

**Comment by jwakely** [comment]
@tlanc007 I don't see any benefit to your suggested text.

> A static class function exists outside the class and can be used without a properly instantiated object of that class. 

It can be used without any object. Period. "Properly instantiated" has nothing to do with it, and sounds ominous, as though static member functions allow something unsafe.

> One such use is to overcome restrictions with constructors. 

I agree that factory functions can be a valid use of static member functions, but not because of restrictions of constructors.

@cubbimew maybe it belongs there, although I can think of a few cases where static member functions are useful that don't apply to non-static member functions.

C.4 does apply sometimes, so maybe we could add:

This rule applies to static member functions as well as non-static member functions.
For example, a generator function that creates objects using public constructors
doesn't need special access to the class, and can be a non-member function (e.g. `std::make_pair`).
A generator function that creates objects using private constructors (or other non-public state)
needs to be a friend or a member (e.g. `std::locale::global(const locale&)`).

(Although I'm not sure `std::locale` is a great exemplar for anything :-)

But there are times when static member functions are useful that have no equivalent for non-static member functions:

It sometimes makes sense for functions that are part of the interface of a class to be
static member functions if they take no arguments, for example `std::chrono::system_clock::now()`.
Because you don't need an instance of `system_clock` to call this function, and it takes no arguments,
there would be no way to distinguish `system_clock::now()` from `some_other_clock::now()`.
By making them static member functions the different `now()` functions are scoped to the clock
type they belong to, and it's clear which clock's `now()` you are calling.

A related case it for traits classes, where a family of related functions are grouped together as static member functions of a class, or class template. This allows the trait type to be provided as a template argument to some generic API which needs the family of related functions. For example, `std::basic_string` has a traits type as a template parameter, and never creates an instance of the traits type, but makes use of the static member functions belonging to that type. If they were non-member functions then `basic_string` would need lots more template parameters so each `char_traits` operation could be specified as an argument. Grouping the related functions as static member functions also means that a specialization of a class template like `char_traits` or `numeric_limits` defines all of the functions in one place, instead of having to specialize a large number of non-member functions.

**Comment by cubbimew** [comment]
@jwakely indeed, [steady_clock::now](http://en.cppreference.com/w/cpp/chrono/steady_clock/now) and [char_traits::length](http://en.cppreference.com/w/cpp/string/char_traits/length) are very proper uses of static member functions ([numeric_limits::max](http://en.cppreference.com/w/cpp/types/numeric_limits/max) not so much anymore - we have variable templates). Sounds like a guideline in the making!

**Comment by viboes** [comment]
I agree with the use of static member functions/variables when this is a configuration point. I see ```Clock::now``` as a configuration point.

However the guideline should include a solution. Where the non-member functions should be located. Some times, as it is the case for factories, it seems reasonable to locate them in the same namespace as the built type. Sometimes these functions are related to some specific domain or work for a specific Concept. Where these related non-member functions should be located? In a specific namespace?
